src/rust/engine/fs/store/src/lib.rs

```diff
@@ -238,29 +238,51 @@ impl RemoteStore {
 
   /// Download the digest to the local byte store from this remote store. The function `f_remote`
   /// can be used to validate the bytes.
-  async fn download_digest_to_local<
-    FRemote: Fn(Bytes) -> Result<(), String> + Send + Sync + 'static,
-  >(
+  async fn download_digest_to_local(
     &self,
     local_store: local::ByteStore,
     digest: Digest,
     entry_type: EntryType,
-    f_remote: FRemote,
+    f_remote: Option<&(dyn Fn(Bytes) -> Result<(), String> + Send + Sync + 'static)>,
   ) -> Result<(), StoreError> {
     let remote_store = self.store.clone();
     self
       .maybe_download(digest, async move {
-        let bytes = remote_store.load_bytes(digest).await?.ok_or_else(|| {
-          StoreError::MissingDigest(
-            "Was not present in either the local or remote store".to_owned(),
-            digest,
-          )
-        })?;
+        // if there's a function to call, always just buffer fully into memory
```

Comment:
Would be good to put this behavior in the function rustdoc, so that it's clear that validation functions should only be used for small content.

Also, for comments, we try to use complete sentences with periods: see https://www.pantsbuild.org/docs/style-guide#comments